design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+# Improvements to virtual chunks handling
+
+## What we need
+
+1. Virtual chunks can be stored in all supported object stores
+2. Multiple credentials sets for virtual chunk resolution
+3. Optimized manifests where the same prefix doesn't need to be repeated for every virtual chunk
+4. Be able to know all the "places" pointed by a repo, without having to load every manifest in every version
+5. Object level change detection (not serving a virtual chunk if the containing file changed)
+
+## Discussion
+
+### Chunks url styles
+
+* Many chunks from a few objects, the prefix can be the file
+  * Example: `s3://some-bucket/some-prefix/some-file.nc` is pointed to by 10k chunks.
+  * Having between 10 and 10k chunks from the same file is not unusual
+* Virtual zarr: each chunk comes from a different object, the prefix is not the object, we still need an extra string
+  * Example: `s3://some-bucket/some-prefix/c/0/0/1` is pointed to by only 1 chunk. But there are many objects like it.
+  * There could be tens of millions of chunks (so different urls)
+
+### Prefixes
+
+Storing the full URL for each chunk duplicates a lot of data in the manifest, making it large. The idea is to try to extract common _prefixes_ and identify them with a short id. Then the manifest only needs to identify the prefix id + maybe an extra string + length + offset.
+
+In the first url style above, the prefix could be the full file, `s3://some-bucket/some-prefix/some-file.nc`. In the second example, there are options for the prefix, it could be `s3://some-bucket/some-prefix/c/0/0/1`, or `s3://some-bucket/some-prefix/c/0/0` or `s3://some-bucket/some-prefix/c/0` or `s3://some-bucket/some-prefix/c`, with different numbers of chunks under it in each case.
+
+Intelligence is needed to optimize what is consider a good prefix. If done badly, we could have:
+
+* as many prefixes as chunks, for example by selecting `s3://some-bucket/some-prefix/c/0/0/1` in the second example above
+* too large of an extra string in each chunk, for example, if we select `s3://some-bucket` as the prefix each chunk needs to add something like `some-prefix/c/0/0`.
+
+### Should prefixes be per snapshot or global?
+
+The advantage of per snapshot is that we don't need to load/parse all the prefixes if the current snapshot uses only a few. A case for this would be a virtual dataset that has been materialized in a newer version. If prefixes are per-snapshot we wouldn't need to load any prefixes.
+
+On the other hand, requirement 4 asks for the full list of "dependencies" for the dataset, and that means for all versions, which requires a global list of prefixes. Without a global list, we would need to fetch the list of prefixes of each snapshot.
```

Comment:
I think we should sharpen the requirement about what exactly these dependencies are. Does the prefix have to be part of it? That feels like a bit of a separate concern. The point of the dependencies is to know what credentials will be required and where on the internet the virtual chunks might live. For these purposes, I'd suggest it's enough to know

- The platform (e.g. S3)
- The bucket name
- [Optional] The region
- [Optional] The endpoint URL

Reply:
I think this is much more clear in my latest commit.